### PR TITLE
FloodFill (#25)

### PR DESCRIPTION
**Pull request by defart**
Flood Fill 

* Added Stack struct in the util package
* Added paint package and a FloodFill implementation
* Added PhpStorm project folder to .gitignore

### CODE REVIEW

**Comment by anthonynsimon** [comment]
Looks good so far, thanks for adding tests one time. Will review it in a second.

File: paint/fill.go [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+}
+
+func isColorMatch(im *image.RGBA, pixel int, matchColor color.Color, fuzzSquared float64) bool {
+
+	i := pixel
+	c1 := color.NRGBAModel.Convert(matchColor).(color.NRGBA)
+	c2 := color.NRGBA{R: im.Pix[i+0], G: im.Pix[i+1], B: im.Pix[i+2], A: im.Pix[i+3]}
+
+	alphaDiff := float64(c1.A) - float64(c2.A)
+	distanceR := math.Max(math.Pow(float64(c1.R)-float64(c2.R), 2), math.Pow(float64(c1.R)-float64(c2.R)-alphaDiff, 2))
```
Comment (anthonynsimon) [inline paint/fill.go]: Cache values to reduce casting and statement length.

File: paint/fill.go [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+	"errors"
+	"github.com/anthonynsimon/bild/clone"
+	"github.com/anthonynsimon/bild/util"
+	"image"
+	"image/color"
+	"math"
+)
+
+const (
+	MaxDistance = 510.0
```
Comment (defart) [inline paint/fill.go]: Fixed, unexported

File: paint/fill.go [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+)
+
+type Point struct {
```
Comment (defart) [inline paint/fill.go]: Fixed, unexported

File: paint/fill.go [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+	X, Y                  int
+	MarkedFromBelow       bool
+	MarkedFromAbove       bool
+	PreviousFillEdgeLeft  int
+	PreviousFillEdgeRight int
+}
+
+//Starts a flood fill of fillColor in the image with a start point and a defined fuzz (percentage of max color distance tolerated)
```
Comment (defart) [inline paint/fill.go]: Fixed

File: paint/fill.go [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+//Returns filled image and error
+func FloodFill(img image.Image, startingPoint image.Point, fillColor color.Color, fuzz float64) (*image.RGBA, error) {
```
Comment (defart) [inline paint/fill.go]: Shortened the param names

File: paint/fill.go [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+
+	var st util.Stack
+	var point Point
+	visited := make(map[int]bool)
+	im := clone.AsRGBA(img)
+
+	maxX := im.Bounds().Dx() - 1
+	maxY := im.Bounds().Dy() - 1
+	if startingPoint.X > maxX || startingPoint.X < 0 || startingPoint.Y > maxY || startingPoint.Y < 0 {
+		return nil, errors.New("Starting point not in bounds!")
```
Comment (defart) [inline paint/fill.go]: Returning just *image.RGBA

File: paint/fill.go (hunk 1 above)
Comment (defart) [inline paint/fill.go]: Fixed

File: util/stack.go [hunk 7]
```diff
@@ -0,0 +1,30 @@
+package util
+
+type Stack struct {
```
Comment (defart) [inline util/stack.go]: Added doc

File: util/stack.go [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+	top  *StackElement
+	size int
+}
+
+type StackElement struct {
```
Comment (defart) [inline util/stack.go]: Fixed, unexported

File: paint/fill_test.go [hunk 9]
```diff
@@ -0,0 +1,72 @@
+package paint
+
+import (
+	"github.com/anthonynsimon/bild/util"
+	"image"
+	"image/color"
+	"testing"
+)
+
+func TestFloodFIll(t *testing.T) {
```
Comment (defart) [inline paint/fill_test.go]: Fixed

File: util/stack_test.go [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+	if intElement != pop1 {
+		t.Fatalf("Expected element to be %d, got %d", intElement, pop1)
+	}
+	if stringElement != pop2 {
+		t.Fatalf("Expected element to be %s, got %s", stringElement, pop2)
+	}
+	if floatElement != pop3 {
+		t.Fatalf("Expected element to be %6.2f, got %6.2f", floatElement, pop3)
+	}
+	if true != noElementsLeft {
```
Comment (defart) [inline util/stack_test.go]: Fixed

File: paint/fill.go [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+				} else {
+					break
+				}
+			}
+			rightFillEdge := xpos + 1
+			if rightFillEdge > maxX {
+				rightFillEdge = maxX
+			}
+
+			//skip every second check for pixels above and below
```
Comment (defart) [inline paint/fill.go]: I've added the no skip version, and made two benchmarks. The skip version is ~30% faster.
The additional branching i think beats the pushing to the stack and checking has the point been already visited for almost 50% more points. It's your call which one we keep, i am more for the version with the column skip.

Comment (anthonynsimon) [inline paint/fill.go]: Ok, then we definitely keep the column skip.

**Pull request merged**